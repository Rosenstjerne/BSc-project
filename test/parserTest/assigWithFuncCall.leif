function int multiply(int x, int y) {
  return x * y;
}

var int a, b, result;
a = 4;
b = 6;
result = multiply(a, b);
print(result);
