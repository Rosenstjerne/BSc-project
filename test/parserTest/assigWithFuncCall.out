
.data
mmapEr:
                .string "mmap failed\n" # error message string for mmap

form:
                .string "%d\n"          # form string for C printf

formTrue:
                .string "true\n"        # form string for C printf

formFalse:
                .string "false\n"       # form string for C printf

.text

.globl main

main:

                                        # CALLEE PROLOGUE
                pushq %rbp              # save caller's base pointer
                movq %rsp, %rbp         # make stack pointer new base pointer

                addq $-24, %rsp         # allocate space for local variables


                movq %rsp, %rbx         # Saves stack pointer in %rbx
                addq $-32, %rsp         # allocate space for extra intermediate
                                        # stack memory

                pushq %r8               # %r8 is callee save
                pushq %r9               # %r9 is callee save
                pushq %r10              # %r10 is callee save
                pushq %r11              # %r11 is callee save
                pushq %r12              # %r12 is callee save
                pushq %r13              # %r13 is callee save
                pushq %r14              # %r14 is callee save
                pushq %r15              # %r15 is callee save

                movq $0, -24(%rbp)      # Initializes result to 0
                movq $0, -16(%rbp)      # Initializes b to 0
                movq $0, -8(%rbp)       # Initializes a to 0
                movq $4, %r12           # Moves integer into reg4
                movq %rbp, %rdx         # preparing for static link computation
                movq %r12, -8(%rdx)     # Move param reg4 into ['fun_1_main',
                                        # 'var_0_a'] (a)
                movq $6, %r14           # Moves integer into reg6
                movq %rbp, %rdx         # preparing for static link computation
                movq %r14, -16(%rdx)    # Move param reg6 into ['fun_1_main',
                                        # 'var_1_b'] (b)

                pushq %r8               # %r8 is callee save
                pushq %r9               # %r9 is callee save
                pushq %r10              # %r10 is callee save
                pushq %r11              # %r11 is callee save
                pushq %r12              # %r12 is callee save
                pushq %r13              # %r13 is callee save
                pushq %r14              # %r14 is callee save
                pushq %r15              # %r15 is callee save
                pushq -24(%rbx)         # stack reg save
                pushq -16(%rbx)         # stack reg save
                pushq -8(%rbx)          # stack reg save
                pushq 0(%rbx)           # stack reg save

                movq %rbp, %rdx         # preparing for static link computation
                movq -8(%rdx), %rcx     # 
                movq %rcx, 0(%rbx)      # Move param var_0_a (a) into reg8
                movq %rbp, %rdx         # preparing for static link computation
                movq -16(%rdx), %rcx    # 
                movq %rcx, -8(%rbx)     # Move param var_1_b (b) into reg9
                pushq -8(%rbx)          # push arguement to stack
                pushq 0(%rbx)           # push arguement to stack
                pushq %rbp              # set up static link for inner function
                callq fun_2_multiply    # 
                popq %rcx               # Pops parent base pointer from stack
                popq %rcx               # Pops argument from stack
                popq %rcx               # Pops argument from stack

                popq 0(%rbx)            # stack reg save
                popq -8(%rbx)           # stack reg save
                popq -16(%rbx)          # stack reg save
                popq -24(%rbx)          # stack reg save
                popq %r15               # restore callee save register %r15
                popq %r14               # restore callee save register %r14
                popq %r13               # restore callee save register %r13
                popq %r12               # restore callee save register %r12
                popq %r11               # restore callee save register %r11
                popq %r10               # restore callee save register %r10
                popq %r9                # restore callee save register %r9
                popq %r8                # restore callee save register %r8

                movq %rax, -16(%rbx)    # Moves the return value from %rax into
                                        # reg10
                movq %rbp, %rdx         # preparing for static link computation
                movq -16(%rbx), %rcx    # 
                movq %rcx, -24(%rdx)    # Move param reg10 into ['fun_1_main',
                                        # 'var_2_result'] (result)
                movq %rbp, %rdx         # preparing for static link computation
                movq -24(%rdx), %rcx    # 
                movq %rcx, -24(%rbx)    # Move param var_2_result (result) into
                                        # reg11
                movq -24(%rbx), %rcx    # Moves reg11 out of the stack for
                                        # print
                                        # PRINTING
                leaq form(%rip), %rdi   # pass 1. argument in %rdi
                movq %rcx, %rsi         # Moves printable object to rsi
                xorq %rax, %rax         # No floating point arguments

                callq printf@plt        # calls the printf method
                                        # END OF PRINTING
end_main:

                popq %r15               # restore callee save register %r15
                popq %r14               # restore callee save register %r14
                popq %r13               # restore callee save register %r13
                popq %r12               # restore callee save register %r12
                popq %r11               # restore callee save register %r11
                popq %r10               # restore callee save register %r10
                popq %r9                # restore callee save register %r9
                popq %r8                # restore callee save register %r8

                addq $32, %rsp          # deallocate stack space for extra
                                        # intermediate stack memory


                                        # CALLEE EPILOGUE
                movq %rbp, %rsp         # restore stack pointer
                popq %rbp               # restore base pointer
                ret                     # return from call

fun_2_multiply:

                                        # CALLEE PROLOGUE
                pushq %rbp              # save caller's base pointer
                movq %rsp, %rbp         # make stack pointer new base pointer

                addq $0, %rsp           # allocate space for local variables

                movq %rbp, %rdx         # preparing for static link computation
                movq 24(%rdx), %r8      # Move param param_3_x (x) into reg0
                movq %rbp, %rdx         # preparing for static link computation
                movq 32(%rdx), %r9      # Move param param_4_y (y) into reg1
                movq %r9, %rcx          # Moves reg1 out into %rcx for
                                        # operation
                movq %r8, %rax          # Moves reg0 out into %rax for
                                        # operation
                imulq %rcx, %rax        # Operation: reg0 * reg2
                movq %rax, %r10         # Moves result into reg2
                movq %r10, %rax         # Moves reg2 out into %rax for return
end_fun_2_multiply:

                                        # CALLEE EPILOGUE
                movq %rbp, %rsp         # restore stack pointer
                popq %rbp               # restore base pointer
                ret                     # return from call


mmap_error:
                leaq mmapEr(%rip), %rdi # pass 1. arg in %rdi
                xor %rax, %rax          # no floating point arguments
                callq printf            # calls printf method

                movq $60, %rax          # prepare for exit syscall
                xor %rdi, %rdi          # exit status 0
                syscall                 # exit


